core/src/main/scala/org/apache/spark/internal/io/FileCommitProtocol.scala

```diff
@@ -100,9 +101,34 @@ abstract class FileCommitProtocol extends Logging {
    * if a task is going to write out multiple files to the same dir. The file commit protocol only
    * guarantees that files written by different tasks will not conflict.
    */
+  @deprecated("use newTaskFile", "3.2.0")
   def newTaskTempFileAbsPath(
       taskContext: TaskAttemptContext, absoluteDir: String, ext: String): String
 
+  /**
+   * Notifies the commit protocol that a new file is added. Must be called on the executors when
+   * running tasks.
+   *
+   * The "stagingPath" parameter is the current path of new file. The "finalPath" parameter if
+   * specified, is the final path of file. The "finalPath" parameter is optional here because
+   * caller can leave up to file commit protocol to decide the final path. The "stagingDir"
+   * parameter if specified, is the sub-directory used to specify dynamic partitioning. The
+   * "stagingDir" parameter is optional here for non-dynamic partitioning.
+   *
+   * Important: it is the caller's responsibility to add uniquely identifying content to
+   * `stagingPath` and `finalPath`. The file commit protocol only guarantees that files written by
+   * different tasks will not conflict. This API should be preferred to use instead of deprecated
+   * [[newTaskTempFile]] and [[newTaskTempFileAbsPath]].
+   */
+  def newTaskFile(
+      taskContext: TaskAttemptContext,
+      stagingPath: String,
+      finalPath: Option[String],
+      stagingDir: Option[String]): Unit = {
+    // No-op as default implementation to be backward compatible with custom [[FileCommitProtocol]]
+    // implementations before Spark 3.2.0.
```

Comment:
I'd like to understand more about backward compatibility. If there is an old impl of file commit protocol, in Spark 3.2 the impl will commit nothing, and this silent "noop" is even worse than an explicit compile error.

This is a developer API, if there is really no way to keep backward compatibility, let's just break the API and force people to correct their impl according to the new API.

Reply:
@cloud-fan - I think it's impossible for me to keep backward compatibility cleanly. Updated here to just introduce new method and break the API.